spark2/src/main/java/org/apache/iceberg/spark/source/Reader.java

```diff
@@ -385,16 +387,40 @@ private static void mergeIcebergHadoopConfs(
         }
       }
 
-      try (CloseableIterable<CombinedScanTask> tasksIterable = scan.planTasks()) {
-        this.tasks = Lists.newArrayList(tasksIterable);
-      } catch (IOException e) {
-        throw new RuntimeIOException(e, "Failed to close table scan: %s", scan);
-      }
+      this.tasks = planScan(scan);
     }
-
     return tasks;
   }
 
+  private List<CombinedScanTask> planScan(TableScan scan) {
+    // TODO Need to only use distributed planner for supported implementations and add some heuristics about when
+    //  to use
+    if (scan instanceof DataTableScan) {
+      return planDistributedScan(scan);
+    } else {
+      return planLocalScan(scan);
+    }
+  }
+
+  private List<CombinedScanTask> planDistributedScan(TableScan scan) {
+    List<CombinedScanTask> result;
+    try {
```

Comment:
I think ideally we attempt to do both a local and distributed planning in parallel and just use whichever returns first but this could be a future improvement.